flow/display_list.cc

```diff
@@ -0,0 +1,1248 @@
+  const SkBlendMode mode;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawColor(color, mode);
+  }
+};
+
+// The common data is a 4 byte header with an unused 4 bytes
+// SkRect is 16 more bytes, using 20 bytes which rounds up to 24 bytes total
+//        (4 bytes unused)
+// SkRRect is 52 more bytes, which packs efficiently into 56 bytes total
+// SkPath is 16 more bytes, using 20 bytes which rounds up to 24 bytes total
+//        (4 bytes unused)
+#define DEFINE_DRAW_1ARG_OP(op_name, arg_type, arg_name)         \
+  struct Draw##op_name##Op final : DLOp {                        \
+    static const auto kType = DisplayListOpType::Draw##op_name;  \
+                                                                 \
+    Draw##op_name##Op(arg_type arg_name) : arg_name(arg_name) {} \
+                                                                 \
+    const arg_type arg_name;                                     \
+                                                                 \
+    void dispatch(Dispatcher& dispatcher) const {                \
+      dispatcher.draw##op_name(arg_name);                        \
+    }                                                            \
+  };
+DEFINE_DRAW_1ARG_OP(Rect, SkRect, rect)
+DEFINE_DRAW_1ARG_OP(Oval, SkRect, oval)
+DEFINE_DRAW_1ARG_OP(RRect, SkRRect, rrect)
+DEFINE_DRAW_1ARG_OP(Path, SkPath, path)
+#undef DEFINE_DRAW_1ARG_OP
+
+// The common data is a 4 byte header with an unused 4 bytes
+// 2 x SkPoint is 16 more bytes, using 20 bytes rounding up to 24 bytes total
+//             (4 bytes unused)
+// SkPoint + SkScalar is 12 more bytes, packing efficiently into 16 bytes total
+// 2 x SkRRect is 104 more bytes, using 108 and rounding up to 112 bytes total
+//             (4 bytes unused)
+#define DEFINE_DRAW_2ARG_OP(op_name, type1, name1, type2, name2) \
+  struct Draw##op_name##Op final : DLOp {                        \
+    static const auto kType = DisplayListOpType::Draw##op_name;  \
+                                                                 \
+    Draw##op_name##Op(type1 name1, type2 name2)                  \
+        : name1(name1), name2(name2) {}                          \
+                                                                 \
+    const type1 name1;                                           \
+    const type2 name2;                                           \
+                                                                 \
+    void dispatch(Dispatcher& dispatcher) const {                \
+      dispatcher.draw##op_name(name1, name2);                    \
+    }                                                            \
+  };
+DEFINE_DRAW_2ARG_OP(Line, SkPoint, p0, SkPoint, p1)
+DEFINE_DRAW_2ARG_OP(Circle, SkPoint, center, SkScalar, radius)
+DEFINE_DRAW_2ARG_OP(DRRect, SkRRect, outer, SkRRect, inner)
+#undef DEFINE_DRAW_2ARG_OP
+
+// 4 byte header + 28 byte payload packs efficiently into 32 bytes
+struct DrawArcOp final : DLOp {
```

Comment:
Full disclosure, I stopped double checking each struct definition right about here as it is just repetition of the same pattern.

Perhaps (later if needed) this entire file can be generated from a manifest? It's fairly straightforward to make a target depend on a generated TU in GN. It will also allow us to switch implementations on a whim. For instance, to see if Dan's concerns make sense on whether we should just use vtable dispatch.

Even if I am unsure if we could have just gotten away with simple vtable dispatch, I couldn't spot any obvious deficiencies with this approach.

Reply:
I like the idea for generating everything from a manifest as that would improve clarity on everything. I'll look into that as a follow-on along with investigating different storage solutions as @dnfield was suggesting.